wcfsetup/install/files/lib/system/image/adapter/GDImageAdapter.class.php

```diff
@@ -493,6 +493,47 @@ public function overlayImageRelative($file, $position, $margin, $opacity) {
 		// does nothing
 	}
 	
+	/**
+	 * @inheritDoc
+	 */
+	public function saveImageAs($image, string $filename, string $type, int $quality = 100): void {
+		if (!$this->isImage($image)) {
+			throw new SystemException("Given image is not a valid image resource.");
+		}
+		
+		ob_start();
+		
+		imagealphablending($image, false);
+		imagesavealpha($image, true);
+		
+		switch ($type) {
+			case "gif":
+				imagegif($image);
+				break;
+				
+			case "jpg":
+			case "jpeg":
+				imagejpeg($image, null, $quality);
+				break;
+				
+			case "png":
+				imagepng($image, null, $quality);
+				break;
+			
+			case "webp":
+				imagewebp($image, null, $quality);
+				break;
+			
+			default:
+				throw new \LogicException("Unreachable");
```

Comment:
This is not _actually_ unreachable (because code might call this method directly), use an `InvalidArgumentException`.

Reply:
I disagree on that one. The adapters expect the validation to have happened in `ImageAdapter` which is the proper wrapper.

It is common for methods to have their input pre-validated by the wrapper and acting directly on the input without further validation. Yes, you can circumvent this wrapper by instantiiating the actual implementation, but at that point you are bypassing the intended API.

I opted for the `LogicException` because we commonly use them for case which should not happen and are the consequence of an actual error in the program flow. One such error would be the support for more file types in the wrapper without implementing them in the actual image handler.

Reply:
FWIW: You appear to have adjusted *one* of the two Unreachable exceptions: https://github.com/WoltLab/WCF/commit/96fcf43c58aa8f87af1be30f11f5bb090fa13fcc#diff-e14386efab40bca7cfcc86fe25cb9ac966320587154c2eabd717b10ee118c120R533